servlet.py

```diff
@@ -587,20 +594,107 @@ def urlRepl(self, base, attr, quote, aurl):
             newurl = aurl
         return ' {a}={q}{u}{q}'.format(a=attr, u=newurl, q=quote)
 
-    def cleanHtml(self, html):
-        return html.replace('&shy;', '').replace('­', '')  # literal and entity soft hyphen
-
-    def htmlToText(self, html, url):
+    def unescape(self, page):
+        # First workaround old bug that exists in a lot of
+        # Windows-based web pages, see
+        # http://stackoverflow.com/a/1398921/69663 :
+        page = page.replace('&#150;', '&#8211;')
+        # Unescape all other entities the regular way:
+        return html.unescape(page)
+
+    def cleanHtml(self, page, urlbase):
+        page = self.unescape(page)
+        if urlbase.netloc in ['www.avvir.no', 'avvir.no']:
+            page = re.sub(r'([a-zæøåášžđŋ])=([a-zæøåášžđŋ])',
+                          '\\1\\2',
+                          page)
+        return page.replace('­', '')  # literal and entity soft hyphen
+
+    def htmlToText(self, page, url):
+        encoding = "utf-8"
         if chardet:
-            encoding = chardet.detect(html).get("encoding", "utf-8")
-        else:
-            encoding = "utf-8"
+            encoding = chardet.detect(page).get("encoding", "utf-8") or encoding
         base = urlparse(url)
-        text = html.decode(encoding)
+        text = self.cleanHtml(page.decode(encoding), base)  # type: str
         return re.sub(r' (href|src)=([\'"])(..*?)\2',
                       lambda m: self.urlRepl(base, m.group(1), m.group(2), m.group(3)),
                       text)
 
+    def setCached(self, pair, url, translated, origtext):
+        """Cache translated text for a pair and url to memory, and disk.
+        Also caches origtext to disk; see cachePath."""
+        if pair not in self.url_cache:
+            self.url_cache[pair] = {}
+        elif len(self.url_cache[pair]) > self.max_inmemory_url_cache:
+            self.url_cache[pair] = {}
+        ts = time.strftime('%a, %d %b %Y %H:%M:%S GMT', time.gmtime(time.time()))
+        self.url_cache[pair][url] = (ts, translated)
+        if self.url_cache_path is None:
+            logging.info("No --url-cache-path, not storing cached url to disk")
```

Comment:
@unhammer I'm assuming that this code came from your branch. Why not just use https://docs.python.org/3/library/tempfile.html?